### PR TITLE
make debug build on Windows MSVC compile

### PR DESCRIPTION
**Pull request by jgehw**
- fix missing-return-statement errors
- add partial dummy template specializations (which are not needed but without optimization the compiler doesn't find out about it)

The second part is a supplement to PR #42307.

### CODE REVIEW

**Review by mihaimaruseac — COMMENTED** [review]
`LOG(FATAL)` calls `std::abort`. So `return` after it should not be needed. Is there a compile failure causing these returns to be needed? If yes, can we actually try to use `std::abort`?

**Comment by jgehw** [comment]
> `LOG(FATAL)` calls `std::abort`. So `return` after it should not be needed. Is there a compile failure causing these returns to be needed? If yes, can we actually try to use `std::abort`?

yes, on Windows debug MSVC (2019) build I get errors like these:
`C:\...\execroot\org_tensorflow\tensorflow\core\framework\device_base.cc(72) : error C4716: 'tensorflow::DeviceBase::name': must return a value`
`C:\...\execroot\org_tensorflow\tensorflow\core\framework\device_base.cc(68) : error C4716: 'tensorflow::DeviceBase::attributes': must return a value`

Apparently in debug mode MSVC compiler doesn't get that `LOG(FATAL, ...)` will finally call `abort()`...

I agree that using `std::abort()` in all cases, instead of returns, is nicer. I'll push an update to this PR.

File: tensorflow/core/kernels/depthtospace_op.cc [hunk 1]
```diff
@@ -172,6 +172,16 @@ struct DepthToSpaceOpFunctor<CPUDevice, T, FORMAT_NHWC> {
...
+
+#ifdef WIN32
+template <typename T>
+struct DepthToSpaceOpFunctor<CPUDevice, T, FORMAT_NCHW> {
+  void operator()(const CPUDevice& d, typename TTypes<T, 4>::ConstTensor input,
+                  int block_size, typename TTypes<T, 4>::Tensor output) {
+    LOG(FATAL) << "dummy implementation to make debug build compile";
+  }
+};
+#endif
```
Comment (mihaimaruseac) [inline tensorflow/core/kernels/depthtospace_op.cc:184]: Do you have some intuition for why these are needed only on Windows debug build?

File: tensorflow/core/kernels/depthtospace_op.cc [hunk 2]
```diff
@@ -172,6 +172,16 @@ struct DepthToSpaceOpFunctor<CPUDevice, T, FORMAT_NHWC> {
     }
   }
 };
+
+#ifdef WIN32
+template <typename T>
+struct DepthToSpaceOpFunctor<CPUDevice, T, FORMAT_NCHW> {
+  void operator()(const CPUDevice& d, typename TTypes<T, 4>::ConstTensor input,
+                  int block_size, typename TTypes<T, 4>::Tensor output) {
+    LOG(FATAL) << "dummy implementation to make debug build compile";
```
Comment (mihaimaruseac) [inline tensorflow/core/kernels/depthtospace_op.cc:181]: Let's try a wording that is less loaded

```suggestion
    LOG(FATAL) << "Trivial implementation to make debug build compile";
```

**Review by mihaimaruseac — CHANGES_REQUESTED** [review]

File: tensorflow/core/kernels/depthtospace_op.cc (hunk 2 above)
Comment (jgehw) [reply tensorflow/core/kernels/depthtospace_op.cc]: I rephrased the wording as you suggested, here and at the second (symmetric) location it occurred.

File: tensorflow/core/kernels/depthtospace_op.cc (hunk 1 above)
Comment (jgehw) [reply tensorflow/core/kernels/depthtospace_op.cc]: The `if (std::is_same<Device, GPUDevice>::value) {` condition in `DepthToSpaceOp::Compute()` obviously never becomes true for CPU devices. In release mode, compilers optimize out this whole block for CPU device template instantiations. In debug mode, non-Windows compilers seem to "understand" that this branch is dead, too, MSVC not...

**Review by mihaimaruseac — APPROVED** [review]

**Comment by jgehw** [comment]
Windows Bazel CPU CI failed on `cwise_op_gpu.lo.lib(cwise_op_gpu_tanh.cu.obj) : error LNK2005: "unsigned __int64 __cdecl Eigen::internal::_GLOBAL__N__39_cwise_op_gpu_tanh_cu_compute_70_cpp1_ii_ef4b1079::get_random_seed(void)" (?get_random_seed@_GLOBAL__N__39_cwise_op_gpu_tanh_cu_compute_70_cpp1_ii_ef4b1079@internal@Eigen@@YA_KXZ) already defined in cwise_op_gpu.lo.lib(cwise_op_gpu_tanh.cu.obj)`. I don't see a link to this PR's code changes + see the very same error on other PRs. Maybe some other fixes need to be merged into this branch?

**Comment by mihaimaruseac** [comment]
Let's try rerunning

**Comment by mihaimaruseac** [comment]
Things passed, but internally there are a few breakages. Will look into those

File: tensorflow/core/kernels/depthtospace_op.cc (hunk 2 above)
Comment (planetmarshall) [reply tensorflow/core/kernels/depthtospace_op.cc]: Just FYI there's an alternative solution to this using C++ 17 `if constexpr` as mentioned in #42094

**Pull request merged**